Doc/library/itertools.rst

```diff
@@ -464,6 +465,23 @@ loops that truncate the stream.
                       pass
 
 
+.. function:: ilen(iterable)
+
+   Consume the iterator and return the number of consumed elements.
+
+   Returns ``0`` for exhausted iterator.
+
+   Equivalent to::
+
+      def ilen(iterable):
+         length = 0
+         for _ in iterable:
+             length += 1
+         return length
+
```

Comment:
```suggestion
      def ilen(iterable):
         return len(list(iterable))
```

Reply:
`len(list())` is not a good equivalence. One of the main advantages of iterators is that one does not need to hold items in memory. Loop represents implementation much closer and speed does not matter here.

E.g. `islice` rough equivalence. It does not give `iter(list(iterable)[start:stop:sep])`, because it does not represent implementation, which is iteration oriented.

Reply:
Then, `sum(1 for _ in iterable)`.

Reply:
Could maybe do:
```python
for n, _ in enumerate(iterable, 1):
    pass
return n
```

Also thought about
```
sum(1 for _ in iterable)
```

But so far I like most basic loop best.

Reply:
Honestly, I think it's a good practice to also highlight constructions that can be simplified. If you look at the recipes, that's what they do.

Reply:
But recipes are different, that is the whole point of them to make clever use of existing functions.

While these functions are here to represent implementation detail as accurately as possible.

I don't like `sum` because it calls `sum` function, which is not part of implementation.
I also don't like loop too much, because it has `_` in it, which doesn't look good.

Reply:
Well.. I think we should wait for another opinion on that one.

Reply:
I will go with `sum`. It does represent accurately enough and has a flavour of "it is as simple as". Loop is a bit of an overkill in verbosity here.

Reply:
I think the for loop was fine. This looks like an optimisation, but isn't even one:

```shell
$ python -m timeit "length = 0" "for _ in range(100): length += 1"  
200000 loops, best of 5: 1.56 usec per loop
$ python -m timeit "length = sum(1 for _ in range(100))"  
100000 loops, best of 5: 3.34 usec per loop
```

The only other one-liner in the documentation is this:

```python
def compress(data, selectors):
    return (datum for datum, selector in zip(data, selectors) if selector)
```

And I would argue that this is more readable:

```python
def compress(data, selectors):
    for datum, selector in zip(data, selectors):
        if selector:
            yield datum
```

Reply:
I agree with you, `generator function` is much more friendly to my eyes.

If `try-exept` wasn't so cumbersome I would go with `while-try-except`, but it looks scary.

Switching back to `for` loop. It is the simplest after all.